File: Julia/textbook_track/chapter01/ba1b.jl
# Find the Most Frequent Words in a String

function most_frequent_k_mers(text, k)
    counts = Dict{AbstractString, Int64}()
    for ix in 1:(length(text)-k+1)
        k_mer = text[ix:ix+k-1]
        count = get(counts, k_mer, 0)
        counts[k_mer] = count + 1
    end
    maxcount = maximum(values(counts))
    [k_mer for (k_mer, count) in counts if count == maxcount]
end


function main()
    text = readline()
    k = parse(Int64, readline())
    result = most_frequent_k_mers(text, k)
    println(join(result, " "))
end


main()

File: OCaml/src/textbook_track/chapter01/ba1b.ml
let most_frequent_k_mers (text : string) (k : int) : string list =
    let n = String.length text in
    let counter = Hashtbl.create (n - k) in
    let get_or_else table key default =
        match Hashtbl.find_opt table key with
            | None -> default
            | Some x -> x in
    let update_counter k_mer =
        let count = get_or_else counter k_mer 0 in
        Hashtbl.replace counter k_mer (count + 1) in
    let indices = Seq.init (n - k) (fun ix -> ix) in
    Seq.iter (fun ix -> update_counter (String.sub text ix k)) indices;
    let max_count = Hashtbl.fold (fun _ count acc -> max count acc) counter 0 in
    Hashtbl.fold (fun k_mer count acc -> if count = max_count then k_mer :: acc else acc) counter []


let () =
    let text = read_line () in
    let k = read_int () in
    let result = most_frequent_k_mers text k in
    result |> String.concat " " |> print_endline
